gclient/src/api/calls.rs

```diff
@@ -109,27 +146,78 @@ impl GearApi {
 
     /// `pallet_gear::claim_value`
     pub async fn claim_value(&self, message_id: MessageId) -> Result<(u128, H256)> {
-        let data = self
+        let value = self
             .get_from_mailbox(self.0.account_id().clone(), message_id)
-            .await?;
+            .await?
+            .map(|(message, _interval)| message.value());
 
         let tx = self.0.claim_value(message_id).await?;
 
-        let events = tx.wait_for_success().await?;
+        for event in tx.wait_for_success().await?.iter() {
+            if let Event::Gear(GearEvent::UserMessageRead { .. }) = event?.event {
+                return Ok((
+                    value.expect("Data appearance guaranteed above"),
+                    tx.block_hash(),
+                ));
+            }
+        }
 
-        let (message, _interval) = data.expect("Data appearance guaranteed above");
+        Err(Error::EventNotFound)
+    }
 
-        for event in events.iter() {
-            if let Event::Gear(GearEvent::UserMessageRead { id, reason }) = event?.event {
-                if MessageId::from(id) == message.id()
-                    && reason == Reason::Runtime(UserMessageReadRuntimeReason::MessageClaimed)
-                {
-                    return Ok((message.value(), tx.block_hash()));
-                }
+    /// `pallet_gear::claim_value` batched.
+    pub async fn claim_value_batch(
+        &self,
+        args: impl IntoIterator<Item = MessageId>,
+    ) -> Result<(Vec<Result<u128>>, H256)> {
+        let message_ids: Vec<_> = args.into_iter().collect();
+
+        let calls: Vec<_> = message_ids
+            .clone()
+            .into_iter()
+            .map(|message_id| {
+                RuntimeCall::Gear(GearCall::claim_value {
+                    message_id: message_id.into(),
+                })
+            })
+            .collect();
+
+        let mut values = BTreeMap::new();
+
+        for message_id in message_ids {
+            values.insert(
+                message_id,
+                self.get_from_mailbox(self.0.account_id().clone(), message_id)
+                    .await?
+                    .map(|(message, _interval)| message.value()),
+            );
+        }
```

Comment:
Consider avoiding excessive `clone` and `collect`.
```suggestion
        let mut message_ids = args.into_iter();

        let mut values = BTreeMap::new();

        for message_id in message_ids.by_ref() {
            values.insert(
                message_id,
                self.get_from_mailbox(self.0.account_id().clone(), message_id)
                    .await?
                    .map(|(message, _interval)| message.value()),
            );
        }

        let calls: Vec<_> = message_ids
            .map(|message_id| {
                RuntimeCall::Gear(GearCall::claim_value {
                    message_id: message_id.into(),
                })
            })
            .collect();
```